SOLR-14067 v2 Move Stateless Scripting Update Process to /contrib/scripting module

epugh opened the pull request:
_Once more, with feeling_

# Description

The scripting update processor is both powerful but also potentially dangerous, and shouldn't ship with Solr by default.

# Solution

Move the Stateless Script Update Processor out of `/core` and into `/contrib` so you need to install it.   I have not yet tried to make it a Solr Package.

Also, I am not happy with the word `Stateless` in the name of the update processor.   I'd love to see what others thought on just renaming `StatelessScriptUpdateProcessorFactory` to just `ScriptUpdateProcessorFactory`.   There is not `Stateful` equivalent, and it seems to me to obscure what this update processor is all about!

# Tests

I've run the existing tests and they pass.

# Checklist

Please review the following and check all that apply:

- [ X] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ X] I have created a Jira issue and added the issue ID to my pull request title.
- [ X] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ X] I have developed this patch against the `master` branch.
- [ ] I have run `./gradlew check`.
- [ X] I have added tests for my changes.
- [ ] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

dweiss commented on gradle/documentation/render-javadoc.gradle:
```diff
@@ -278,6 +278,13 @@ configure(project(":solr:contrib:langid")) {
+configure(project(":solr:contrib:scripting-update-processor")) {
+  project.tasks.withType(RenderJavadocTask) {
+    // TODO: clean up split packages
+    javadocMissingIgnore = [ "org.apache.solr.update.processor" ]
```
I would just add the javadoc placeholder instead...

epugh replied:
I think I understood what you mean, and backed out this change..   ;-)

rmuir replied:
can we please change the package name and not create more NEW split package issues!

epugh replied:
Would moving this class to say `org.apache.solr.update.processor.scripting` avoid the split package issue?

epugh replied:
Now under a new `org.apache.solr.scripting.update` package to solve this.

epugh commented:
I put this as Draft as I'm still working on cleaning up refguide.

epugh commented:
Should `ScriptEngineTest` move to the /contrib as well?   It doesn't directly touch the update processor, but I think it exists only because of that code.

epugh commented:
I'm getting there, and I wanted to specifically mention that @chatman other PR was critical for me following the chain of changes required to move this to /contrib.   Thanks @chatman for chasing the all the touchpoints.

epugh commented:
Okay, this PR is kind of "ready".   I've migrated the content out of the Cwiki page, and into the ref guide.   One thing I want to highlight is that I'd like to be able to easily demonstrate the power of the ScriptingUpdateProcessor with the Techproducts example, however to do that, I had to set the `enableStreamBody` to true to make it work.  I hope the fact that it is enabled when you do `bin/solr start -e techproducts` is okay.

epugh commented:
@dweiss I'm hoping you can steer me in the right direction.  I removed the `contrib/scripting-update-processor/src/java/org/apache/solr/update/processor/package-info.java` because Eclipse flagged that there was already one in `core/src/java/org/apache/solr/update/processor/`, and instead added a `package.html`, following another example.   That is making the precommit fail.   Thoughts?

tflobbe commented on solr/contrib/scripting-update-processor/src/java/org/apache/solr/update/processor/StatelessScriptUpdateProcessorFactory.java:
```diff
@@ -0,0 +1,25 @@
+ * @deprecated use {@link ScriptUpdateProcessorFactory} instead.  This class will be removed in an upcoming release.
+ */
+@Deprecated
+public class StatelessScriptUpdateProcessorFactory extends ScriptUpdateProcessorFactory {
```
Just an idea, but you could implement the constructor that logs the deprecation.

epugh replied:
In talking to David, after fighting with the combination of `@deprecated` and no `@Deprecated` and the logging class, we decide to make this a 9.0 breaking change, and not worry about it.  Since it's in a new `scripting` module anyway, under a new path.

tflobbe reviewed (commented):
There are a bunch of unrelated changes here, would you mind cleaning those up?

epugh commented:
I notice that `./gradlew assemble` doesn't include the `scripting-update-processor` in the set of /contrib :-(

dsmiley commented:
I'd love to review but you merged in from master, and I don't know how to get GitHub to only show what you are changing from when you synced, vs the start of the feature branch.  Is there a tip/technique I don't know?  Or you could force push and don't sync yourself up again?

epugh commented:
I am definitly struggling with branches, keeping up with master.   Clearly there is something I don't understand about the workflow with Git.   Plus, I'm very nervous about anything I do..  The idea of doing a "force push" makes me nervous I'm going to mess up everything for everybody.   Makes me think I need to invest time in properly learning how to use Git.  

When I look at folks code, I typically just bring up the files changed view, in this case: https://github.com/apache/lucene-solr/pull/2016/files and then scroll through.  

I wonder if, at this point, the right thing to do is just extract a .patch file, add that back to the ticket, and then apply that to master, and abandon this branch.   Folks have provided lots of great review through this journey on this branch, but now I don't know that we need it anymore.

epugh commented:
I saw the tip in github to add `.diff` and `.patch` to a url to generate those files.  I think I see the issue, in that the `.patch` file generated was 12 MB, and included all the changes that have been dragged in from Master.   Whereas the `.diff` was small, and showed just what files I changed.   For fun, I checked out `master` and did `git apply 2016.diff` and it worked just fine.  It doesn't preserve the fact that I moved a lot of files around, instead they are reflected as separate deletes and creates.

dsmiley commented:
Yes; invest time in learning Git more; it's a prerequisite nowadays.  I'm no Git expert (e.g. not quite Dawid :-) but I know enough, perhaps more than most as I've done some major git surgery on some project splits/mergers and my Stackoverflow answer [here](https://stackoverflow.com/a/17867910/92186) gets upvoted frequently.

I recommend _not_ merging master into the feature branch to keep it up to-date because I have not seen positive experiences with that -- mainly the PR is hard/impossible to review at that point.  I'd love to be proven wrong here!  Just accept that it's not up to date.  But sometimes we  _really_ want to get a PR up to date because we think/know there are changes that likely impact the PR.  In this case, I'd rebase the feature branch from master then force push... but do that sparingly because it kind of somewhat resets the review process, as it's hard to see what you changed when bringing it up to date.  It can be helpful to add a comment to mention where the more interesting conflict resolution points were (if any).  You might do this rebase & force push _after_ the reviews so that you can address non-trivial conflicts from master.  No doubt, when reading about Git, you will hear a lot of  "never" do this/that but it's really more of an "avoid" rather than "never".  If you are afraid to force push, then read about it and try it out.  You're on a fork, so a major screw-up would be contained.  FWIW I do all this stuff from the IntelliJ IDE, which I think is excellent.  I can screen-share with you if if you like (Slack DM me); I'd love to help you out here or really anyone.

epugh commented:
I followed the steps that Joel recommended in another thread, and created a new clean branch, #2215.   I will close this one in favour of that PR which is much more legible.
